File: R/nprobust/R/kdrobust.R
kdrobust <- function(x, eval=NULL, neval=NULL, h=NULL, b=NULL, rho=1, kernel="epa",
                    bwselect=NULL, bwcheck=21, imsegrid=30, level=95, subset = NULL,
                    data = NULL) {

  if (!is.null(data)) {
    mc <- match.call()
    caller_env <- parent.frame()
    .lookup <- function(arg) {
      expr <- mc[[arg]]
      if (is.null(expr)) return(NULL)
      eval(expr, envir = data, enclos = caller_env)
    }
    x <- .lookup("x")
    if ("subset" %in% names(mc)) subset <- .lookup("subset")
  }

  p <- 2
  deriv <- 0
  if (!is.null(subset)) x <- x[subset]

  ## UX prechecks: catch invalid h/b/bwcheck/imsegrid with clear messages
  .nperrs <- character()
  .bwbad <- function(v, name) {
    if (is.null(v))         return(NULL)
    if (!is.numeric(v))     return(paste0(name, " must be numeric."))
    if (any(!is.finite(v))) return(paste0(name, " contains non-finite values (NA/Inf)."))
    if (any(v <= 0))        return(paste0(name, " must be strictly positive."))
    NULL
  }
  .nperrs <- c(.nperrs, .bwbad(h, "h"), .bwbad(b, "b"))
  if (!is.null(bwcheck) && (!is.numeric(bwcheck) || length(bwcheck) != 1 || !is.finite(bwcheck) || bwcheck <= 0))
    .nperrs <- c(.nperrs, "bwcheck must be a single positive finite integer.")
  if (!is.numeric(imsegrid) || length(imsegrid) != 1 || !is.finite(imsegrid) || imsegrid <= 0)
    .nperrs <- c(.nperrs, "imsegrid must be a single positive finite integer.")
  if (!is.numeric(level) || length(level) != 1 || !is.finite(level) || level <= 0 || level >= 100)
    .nperrs <- c(.nperrs, "level must be a single number in (0, 100).")
  if (!is.numeric(rho) || length(rho) != 1 || !is.finite(rho) || rho < 0)
    .nperrs <- c(.nperrs, "rho must be a single non-negative number.")
  if (!is.null(eval)) {
    if (!is.numeric(eval) || any(!is.finite(eval)))
      .nperrs <- c(.nperrs, "eval must be numeric and finite.")
    if (length(eval) == 0L)
      .nperrs <- c(.nperrs, "eval must have at least one element.")
  }
  if (length(.nperrs) > 0) {
    for (.m in .nperrs) warning(.m, call. = FALSE)
    stop("nprobust: invalid input (see warnings above).", call. = FALSE)
  }

  na.ok <- complete.cases(x)
  x <- x[na.ok]
  
  x.min <- min(x);  x.max <- max(x)
  N <- length(x)

  if (!is.null(bwcheck)) {
    if (bwcheck > N) {
      warning("bwcheck (", bwcheck, ") is larger than the sample size (", N,
              "); reducing bwcheck to N.")
      bwcheck <- N
    }
  }

  if (is.null(eval)) {
    if (is.null(neval)) {
      qseq <- seq(0.1,0.9,length.out=30)
      eval <- quantile(x, qseq)
      #eval <- unique(x)
      #qseq <- seq(0,1,1/(20+1))
      #eval <- quantile(x, qseq[2:(length(qseq)-1)])
      #eval <- seq(x.min, x.max, length.out=30)
    }
    else {
      qseq <- seq(0.1,0.9,length.out=neval)
      eval <- quantile(x, qseq)
      #eval <- seq(x.min,x.max,length.out=neval)
      #qseq <- seq(0,1,1/(neval+1))
      #eval <- quantile(x, qseq[2:(length(qseq)-1)])
      #eval <- seq(x.min, x.max, length.out=neval)
    }
  }
  neval <- length(eval)

  ## Precheck (continued): h/b length must be 1 or neval
  .nperrs2 <- character()
  if (!is.null(h) && length(h) != 1L && length(h) != neval)
    .nperrs2 <- c(.nperrs2, paste0("h must have length 1 or neval (=", neval, ")."))
  if (!is.null(b) && length(b) != 1L && length(b) != neval)
    .nperrs2 <- c(.nperrs2, paste0("b must have length 1 or neval (=", neval, ")."))
  if (length(.nperrs2) > 0) {
    for (.m in .nperrs2) warning(.m, call. = FALSE)
    stop("nprobust: invalid input (see warnings above).", call. = FALSE)
  }

  if (is.null(h) & is.null(bwselect) & neval==1) bwselect="mse-dpi"
  if (is.null(h) & is.null(bwselect) & neval>1)  bwselect="imse-dpi"
  
  kernel   <- tolower(kernel)
  bwselect <- tolower(bwselect)

  #####################################################   CHECK ERRORS
  if (!(kernel %in% c("epa","epanechnikov","uni","uniform"))) {
    stop("kernel incorrectly specified. Supported kernels for kdrobust: epa, uni.")
  }
  if (kernel %in% c("epanechnikov")) kernel <- "epa"
  if (kernel %in% c("uniform"))      kernel <- "uni"

  if (identical(bwselect, "all"))
    stop("bwselect=\"all\" is only supported by kdbwselect; kdrobust requires a single method (e.g. \"mse-dpi\", \"imse-dpi\", \"ce-dpi\").",
         call. = FALSE)

  if (!is.null(h)) bwselect <- "Manual"

  kernel.type <- if (kernel == "epa") "Epanechnikov" else "Uniform"

  if (!is.null(h) & rho>0 & is.null(b)) {
    #rho <- rep(1,neval)
    b <- h/rho
  }
  #if (!is.null(h) & !is.null(rho) ) b <- h/rho
  if (is.null(h)) {
      kdbws <- kdbwselect(x=x, eval=eval, bwselect=bwselect, bwcheck=bwcheck, imsegrid=imsegrid, kernel=kernel)
      h <- kdbws$bws[,2]
      b <- kdbws$bws[,3]
      if (rho>0) b <- h/rho
      rho <- h/b  
  }
  
  if (length(h)==1 & neval>1) {
    h <- rep(h,neval)
    b <- rep(b,neval)
    rho   <- h/b
  }

  
  Estimate<-matrix(NA,neval,8)
  colnames(Estimate)<-c("eval","h","b","N","tau.us","tau.bc","se.us","se.rb")
  
  for (i in 1:neval) {
    
    if (!is.null(bwcheck)) {
      bw.min   <- sort(abs(x-eval[i]))[bwcheck]
      h[i]     <- max(h[i], bw.min)
      b[i]     <- max(b[i], bw.min)
      rho[i]   <- h[i]/b[i]
    }
    
    u   <- (x-eval[i])/h[i]
    K.d <- kd.K.fun(u        , v=p,   r=deriv, kernel=kernel)
    L.r <- kd.K.fun(rho[i]*u , v=p+2, r=p,     kernel=kernel)
  
    K     <- K.d$Kx
    M     <- K - rho[i]^(1+p)*L.r$Kx*L.r$k.v
    f.us  <- mean(K)/h[i]
    f.bc  <- mean(M)/h[i]
    se.us <- sqrt((mean((K^2)) - mean(K)^2)/(N*h[i]^2))
    se.rb <- sqrt((mean((M^2)) - mean(M)^2)/(N*h[i]^2))

    if (kernel == "gau") {
      eN <- N
    } else {
      eN <- sum(abs(x - eval[i]) <= max(h[i], b[i]))
    }
    
    Estimate[i,] <- c(eval[i], h[i], b[i], eN, f.us, f.bc, se.us, se.rb) 
  }
  out<-list(Estimate=Estimate, opt=list(p=p, kernel=kernel.type, n=N, neval=neval, bwselect=bwselect))
  out$call <- match.call()
  class(out) <- "kdrobust"
  return(out)
}

print.kdrobust <- function(x,...){
  cat("Call: kdrobust\n\n")

  cat(paste("Sample size (n)                            =     ", x$opt$n,        "\n", sep=""))
  cat(paste("Kernel order for point estimation (p)      =     ", x$opt$p,        "\n", sep=""))
  cat(paste("Kernel function                            =     ", x$opt$kernel,   "\n", sep=""))
  cat(paste("Bandwidth method                           =     ", x$opt$bwselect, "\n", sep=""))
  cat("\n")

  invisible(x)
}

summary.kdrobust <- function(object, alpha = 0.05, sep = 5, ...) {
  z    <- qnorm(1 - alpha / 2)
  CI_l <- object$Estimate[, "tau.bc"] - object$Estimate[, "se.rb"] * z
  CI_r <- object$Estimate[, "tau.bc"] + object$Estimate[, "se.rb"] * z

  out <- list(opt      = object$opt,
              Estimate = object$Estimate,
              alpha    = alpha,
              sep      = sep,
              CI_l     = CI_l,
              CI_r     = CI_r)
  class(out) <- "summary.kdrobust"
  out
}

print.summary.kdrobust <- function(x, ...) {
  cat("Call: kdrobust\n\n")

  cat(paste("Sample size (n)                            =     ", x$opt$n,        "\n", sep=""))
  cat(paste("Kernel order for point estimation (p)      =     ", x$opt$p,        "\n", sep=""))
  cat(paste("Kernel function                            =     ", x$opt$kernel,   "\n", sep=""))
  cat(paste("Bandwidth selection method                 =     ", x$opt$bwselect, "\n", sep=""))
  cat("\n")

  alpha <- x$alpha
  sep   <- x$sep

  cat(paste(rep("=", 14 + 10 + 8 + 10 + 10 + 25), collapse="")); cat("\n")

  cat(format(" ", width= 14 ))
  cat(format(" ", width= 10 ))
  cat(format(" ", width= 8  ))
  cat(format("Point", width= 10, justify="right"))
  cat(format("Std." , width= 10, justify="right"))
  cat(format("Robust B.C.", width=25, justify="centre"))
  cat("\n")

  cat(format("eval"            , width=14, justify="right"))
  cat(format("bw"              , width=10, justify="right"))
  cat(format("Eff.n"           , width=8 , justify="right"))
  cat(format("Est."            , width=10, justify="right"))
  cat(format("Error"           , width=10, justify="right"))
  cat(format(paste("[ ", floor((1-alpha)*100), "%", " C.I. ]", sep="")
             , width=25, justify="centre"))
  cat("\n")

  cat(paste(rep("=", 14 + 10 + 8 + 10 + 10 + 25), collapse="")); cat("\n")

  for (j in 1:nrow(x$Estimate)) {
    cat(format(toString(j), width=4))
    cat(format(sprintf("%3.3f", x$Estimate[j, "eval"]), width=10, justify="right"))
    cat(format(sprintf("%3.3f", x$Estimate[j, "h"])  , width=10, justify="right"))
    cat(format(sprintf("%3.0f", x$Estimate[j, "N"])  , width=8 , justify="right"))
    cat(format(sprintf("%3.3f", x$Estimate[j, "tau.us"]) , width=10, justify="right"))
    cat(format(paste(sprintf("%3.3f", x$Estimate[j, "se.us"]), sep=""), width=10, justify="right"))
    cat(format(paste("[", sprintf("%3.3f", x$CI_l[j]), " , ", sep="")  , width=14, justify="right"))
    cat(format(paste(sprintf("%3.3f", x$CI_r[j]), "]", sep=""), width=11, justify="left"))
    cat("\n")
    if (is.numeric(sep)) if (sep > 0) if (j %% sep == 0) {
      cat(paste(rep("-", 14 + 10 + 8 + 10 + 10 + 25), collapse="")); cat("\n")
    }
  }

  cat(paste(rep("=", 14 + 10 + 8 + 10 + 10 + 25), collapse="")); cat("\n")

  invisible(x)
}

File: Python/nprobust/src/nprobust/kdrobust.py
"""Kernel density with robust bias-corrected inference."""
from __future__ import annotations

import warnings
from dataclasses import dataclass, field
from typing import Any, Dict, Optional

import numpy as np
import pandas as pd
from scipy.stats import norm

from ._kernels import kd_K_fun


_KERNEL_ALIASES = {
    "epa": "epa", "epanechnikov": "epa",
    "uni": "uni", "uniform": "uni",
}
_KERNEL_TYPE_NAMES = {"epa": "Epanechnikov", "uni": "Uniform"}


@dataclass
class KdrobustResult:
    """Container for :func:`kdrobust` results.

    Attributes
    ----------
    Estimate : pandas.DataFrame
        Density estimates and standard errors over the evaluation grid, of
        shape ``(neval, 8)``. Columns: ``eval, h, b, N, tau.us, tau.bc,
        se.us, se.rb``.
    opt : dict
        Options actually used. Keys: ``p, kernel, n, neval, bwselect``.
    call : dict
        Optional dict for storing the call metadata (default empty).
    """
    Estimate: pd.DataFrame
    opt: Dict[str, Any]
    call: Dict[str, Any] = field(default_factory=dict)

    def __repr__(self) -> str:  # pragma: no cover
        return (f"<KdrobustResult n={self.opt['n']} kernel={self.opt['kernel']} "
                f"bwselect={self.opt['bwselect']} neval={self.opt['neval']}>")

    def summary(self, alpha: float = 0.05) -> pd.DataFrame:
        z = norm.ppf(1 - alpha / 2)
        df = self.Estimate.copy()
        df["CI.lower"] = df["tau.bc"] - z * df["se.rb"]
        df["CI.upper"] = df["tau.bc"] + z * df["se.rb"]
        return df[["eval", "h", "b", "N", "tau.us", "se.us",
                   "tau.bc", "se.rb", "CI.lower", "CI.upper"]]

    def tidy(self, conf_level: float = 0.95) -> pd.DataFrame:
        z = norm.ppf(1 - (1 - conf_level) / 2)
        e = self.Estimate
        return pd.DataFrame({
            "eval":      e["eval"].values,
            "estimate":  e["tau.us"].values,
            "std.error": e["se.us"].values,
            "tau.bc":    e["tau.bc"].values,
            "se.rb":     e["se.rb"].values,
            "h":         e["h"].values,
            "b":         e["b"].values,
            "n.eff":     e["N"].values,
            "conf.low":  e["tau.bc"].values - z * e["se.rb"].values,
            "conf.high": e["tau.bc"].values + z * e["se.rb"].values,
        })

    def glance(self) -> pd.DataFrame:
        return pd.DataFrame([{
            "n": self.opt["n"], "neval": self.opt["neval"],
            "kernel": self.opt["kernel"],
            "bwselect": self.opt["bwselect"],
        }])


def _normalize_kernel(kernel: str) -> str:
    k = _KERNEL_ALIASES.get(kernel.lower())
    if k is None:
        raise ValueError(
            "kernel incorrectly specified. Supported kernels for kdrobust: epa, uni."
        )
    return k


def kdrobust(
    x,
    eval=None, neval=None,
    h=None, b=None, rho: float = 1.0,
    kernel: str = "epa",
    bwselect: Optional[str] = None,
    bwcheck: Optional[int] = 21,
    imsegrid: int = 30,
    level: float = 95,
    subset=None,
    data=None,
) -> KdrobustResult:
    """Kernel density estimator with robust bias-corrected inference.

    Implements kernel density point estimators, with robust bias-corrected
    confidence intervals and inference procedures developed in Calonico,
    Cattaneo and Farrell (2018), with related optimality results in Calonico,
    Cattaneo and Farrell (2022). It also implements other estimation and
    inference procedures available in the literature. Companion command:
    :func:`kdbwselect`.

    If ``data`` is a pandas DataFrame, ``x`` and ``subset`` may be passed as
    column-name strings.

    Parameters
    ----------
    x : array-like or column-name string
        Independent variable.
    eval : array-like or None
        Evaluation points. If ``None``, 30 quantile-spaced points (deciles
        0.1 through 0.9 in equal steps) over the support of ``x`` are used.
    neval : int or None
        Number of quantile-spaced evaluation points. Default 30.
    h, b : float, array, or None
        Main bandwidth and bias-correction bandwidth. Each can be scalar or
        have the same length as ``eval``. If not supplied, bandwidths are
        selected by :func:`kdbwselect`.
    rho : float
        Sets ``b = h / rho`` when ``b`` is not supplied. Default 1.
    kernel : {"epa", "uni"}
        Kernel function. Default ``"epa"``.
    bwselect : str or None
        Bandwidth selector. Options are ``"mse-dpi"``, ``"imse-dpi"``,
        ``"imse-rot"``, ``"ce-dpi"``, and ``"ce-rot"``. Use
        :func:`kdbwselect` with ``bwselect="all"`` to report all available
        bandwidth selectors. Default is ``"mse-dpi"`` with one evaluation
        point and ``"imse-dpi"`` otherwise.
    bwcheck : int or None
        If positive, selected bandwidths are enlarged so at least
        ``bwcheck`` observations are available at each evaluation point.
        Default 21.
    imsegrid : int
        Number of evaluation points used to compute IMSE bandwidth selectors.
        Default 30.
    level : float
        Confidence level in percent. Default 95.
    subset : array-like or column-name string, optional
        Boolean rule selecting observations to use.
    data : pandas.DataFrame, optional
        Data source for column-name inputs.

    Returns
    -------
    result : KdrobustResult
        An object with the following attributes:

        - ``Estimate`` (``pandas.DataFrame``, shape ``(neval, 8)``):
          density estimates and standard errors over the evaluation grid.
          Columns: ``eval, h, b, N, tau.us, tau.bc, se.us, se.rb``.
        - ``opt`` (``dict``): options actually used. Keys: ``p, kernel, n,
          neval, bwselect``.
    """
    if data is not None:
        from ._helpers import resolve_name
        x      = resolve_name(x,      data, "x")
        subset = resolve_name(subset, data, "subset")

    x = np.asarray(x, dtype=float).ravel()
    p = 2
    deriv = 0

    if subset is not None:
        idx = np.asarray(subset).ravel()
        if idx.dtype != bool:
            raise TypeError("subset must be a boolean array.")
        x = x[idx]

    x = x[np.isfinite(x)]
    N = len(x)
    if bwcheck is not None and bwcheck > N:
        warnings.warn(f"bwcheck ({bwcheck}) > N ({N}); reducing bwcheck to N.")
        bwcheck = N

    if eval is None:
        if neval is None:
            eval = np.quantile(x, np.linspace(0.1, 0.9, 30))
        else:
            eval = np.quantile(x, np.linspace(0.1, 0.9, int(neval)))
    eval = np.asarray(eval, dtype=float).ravel()
    neval = len(eval)

    if h is None and bwselect is None:
        bwselect = "mse-dpi" if neval == 1 else "imse-dpi"

    kernel = _normalize_kernel(kernel)
    if bwselect is not None: bwselect = bwselect.lower()

    if level <= 0 or level >= 100:
        raise ValueError("level must be in (0, 100).")
    if rho < 0:
        raise ValueError("rho must be non-negative.")

    kernel_type = _KERNEL_TYPE_NAMES[kernel]

    if h is not None:
        h = np.asarray(h, dtype=float).ravel()
        if b is None and rho > 0:
            b = h / rho
        if b is None and rho == 0:
            raise ValueError("When h is provided and rho=0, b must also be provided.")
        b = np.asarray(b, dtype=float).ravel()
        bwselect = "Manual"
    else:
        from .kdbwselect import kdbwselect
        r = kdbwselect(x, eval=eval, bwselect=bwselect, bwcheck=bwcheck,
                       imsegrid=imsegrid, kernel=kernel)
        h = r.bws.iloc[:, 1].to_numpy()
        b = r.bws.iloc[:, 2].to_numpy()
        if rho > 0: b = h / rho

    if len(h) == 1 and neval > 1: h = np.repeat(h, neval); b = np.repeat(b, neval)
    rho_vec = h / b

    ests = np.full((neval, 8), np.nan)
    for i in range(neval):
        h_i = h[i]; b_i = b[i]
        if bwcheck is not None:
            bw_min = np.sort(np.abs(x - eval[i]))[bwcheck - 1]
            h_i = max(h_i, bw_min)
            b_i = max(b_i, bw_min)
            rho_vec[i] = h_i / b_i

        u = (x - eval[i]) / h_i
        K_d = kd_K_fun(u, v=p, r=deriv, kernel=kernel)
        L_r = kd_K_fun(rho_vec[i] * u, v=p + 2, r=p, kernel=kernel)
        K = K_d[0]
        M = K - rho_vec[i] ** (1 + p) * L_r[0] * L_r[1]

        f_us = np.mean(K) / h_i
        f_bc = np.mean(M) / h_i
        se_us = np.sqrt((np.mean(K ** 2) - np.mean(K) ** 2) / (N * h_i ** 2))
        se_rb = np.sqrt((np.mean(M ** 2) - np.mean(M) ** 2) / (N * h_i ** 2))

        bw_eff = max(h_i, b_i)
        eN = int(np.sum(np.abs(x - eval[i]) <= bw_eff))

        ests[i, :] = [eval[i], h_i, b_i, eN, f_us, f_bc, se_us, se_rb]

    est_df = pd.DataFrame(ests, columns=["eval", "h", "b", "N",
                                          "tau.us", "tau.bc", "se.us", "se.rb"])
    opt = dict(p=p, kernel=kernel_type, n=N, neval=neval,
               bwselect=bwselect if bwselect else "Manual")
    return KdrobustResult(Estimate=est_df, opt=opt)
